### PR TITLE
Allow for Q and DQ on hexahedral meshes (except dS integrals)

### PR DESCRIPTION
**Pull request by ksagiyam**
Depends on:

https://github.com/firedrakeproject/tsfc/pull/284

https://github.com/FInAT/FInAT/pull/104

https://github.com/firedrakeproject/fiat/pull/31

### CODE REVIEW

**Comment by ksagiyam** [comment]
@pefarrell This is something we chatted about some time ago...

**Comment by ksagiyam** [comment]
Chatting with @pbrubeck and @pefarrell, I realised that I completely forgot about interior facet integrals of DQ for which quadrature points on `+` and `-` sides must match. I think I should let this PR target Q elements only; I will add some surface integral tests for Q.

**Comment by ksagiyam** [comment]
Disallowed `dS` integrals.

File: firedrake/cython/dmcommon.pyx [hunk 1]
```diff
@@ -325,6 +326,197 @@ def facet_numbering(PETSc.DM plex, kind,
...
+    :arg plex_cone_old: The original DMPlex cone
+    :arg plex_cone_new: The reorderd cone (output)
+
+    This function defines fixed rules to reorder DMPlex cones that facilitates
+    comparison with the associated FIAT cones.
+    The reordered DMPlex cones are later used to construct cell_closure and to
+    compute orientations.
+    """
+    if dm.getCellType(p) == PETSc.DM.PolytopeType.POINT:
+        raise RuntimeError(f"POINT has not cones")
```
Comment (dham) [inline firedrake/cython/dmcommon.pyx:346]: English?

Comment (dham) [reply firedrake/cython/dmcommon.pyx:346]: "POINT has no cone".

File: firedrake/cython/dmcommon.pyx [hunk 2]
```diff
@@ -325,6 +326,197 @@ def facet_numbering(PETSc.DM plex, kind,
...
+    This function defines fixed rules to reorder DMPlex closures
+    for FIAT closures.
+    Constructing cell_closure using _reorder_plex_closure() and
+    reordering plex_cone using _reorder_plex_cone(), we make it
+    sure that the cell orientation is always 0 in entity_orientations().
+    Indeed the same FIAT closure can be obtained merely using
+    _reorder_plex_cone() and ensuring that the cell orientation is 0.
+    """
+    if dm.getCellType(p) == PETSc.DM.PolytopeType.POINT:
+        raise RuntimeError(f"POINT has not cones")
```
Comment (dham) [inline firedrake/cython/dmcommon.pyx:415]: "POINT has no cone"

File: firedrake/mesh.py [hunk 3]
```diff
@@ -962,7 +962,12 @@ def cell_closure(self):
 
             return dmcommon.quadrilateral_closure_ordering(
                 plex, vertex_numbering, cell_numbering, cell_orientations)
-
+        elif cell.cellname() == "hexahedron":
+            # TODO: Should change and use create_cell_closure() for all cell types.
+            import FIAT
```
Comment (dham) [inline firedrake/mesh.py:967]: Move to top of file.

File: tests/output/test_io_function.py [hunk 4]
```diff
@@ -33,6 +33,9 @@ def _get_mesh(cell_type, comm):
     elif cell_type == "quadrilateral":
         mesh = Mesh(join(cwd, "..", "meshes", "unitsquare_unstructured_quadrilaterals.msh"),
                     name=mesh_name, comm=comm)
+    elif cell_type == "hexahedral":
+        mesh = Mesh(join(cwd, "..", "meshes", "cube_hex.msh"),
```
Comment (dham) [inline tests/output/test_io_function.py:37]: Check this is really unstructured.

Possibly also put in a Möbius strip.

**Review by dham — CHANGES_REQUESTED** [review]

File: firedrake/cython/dmcommon.pyx (hunk 1 above)
Comment (ksagiyam) [reply firedrake/cython/dmcommon.pyx]: Fixed.

File: firedrake/cython/dmcommon.pyx (hunk 2 above)
Comment (ksagiyam) [reply firedrake/cython/dmcommon.pyx]: Fixed.

File: firedrake/mesh.py (hunk 3 above)
Comment (ksagiyam) [reply firedrake/mesh.py]: Fixed.

File: tests/output/test_io_function.py (hunk 4 above)
Comment (ksagiyam) [reply tests/output/test_io_function.py:38]: Found that `cube_hex.msh` has a vertex that is shared by 26 cells.

Added `möbius_solid.msh` and added a test in `tests/output/test_io_function.py`.

Screenshot from 2022-11-17 20-06-42

**Review by connorjward — APPROVED** [review]

**Pull request merged**